libraries/lib-dynamic-range-processor/DynamicRangeProcessorUtils.cpp

```diff
@@ -0,0 +1,125 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  DynamicRangeProcessorUtils.cpp
+
+  Matthieu Hodgkinson
+
+**********************************************************************/
+#include "DynamicRangeProcessorUtils.h"
+#include <optional>
+#include <regex>
+#include <stdexcept>
+#include <unordered_map>
+
+namespace DynamicRangeProcessorUtils
+{
+namespace
+{
+
+template <typename Struct> struct SettingDescription
+{
+   double Struct::*const mem {};
+   const char* const key {};
+};
+
+static const std::vector<SettingDescription<CompressorSettings>>
+   compressorSettingDescriptions {
+      { &CompressorSettings::thresholdDb, "thresholdDb" },
+      { &CompressorSettings::makeupGainDb, "makeupGainDb" },
+      { &CompressorSettings::kneeWidthDb, "kneeWidthDb" },
+      { &CompressorSettings::compressionRatio, "compressionRatio" },
+      { &CompressorSettings::lookaheadMs, "lookaheadMs" },
+      { &CompressorSettings::attackMs, "attackMs" },
+      { &CompressorSettings::releaseMs, "releaseMs" },
+      { &CompressorSettings::showInput, "showInput" },
+      { &CompressorSettings::showOutput, "showOutput" },
+      { &CompressorSettings::showActual, "showActual" },
+      { &CompressorSettings::showTarget, "showTarget" },
+   };
+
+const std::vector<SettingDescription<LimiterSettings>>
+   limiterSettingDescriptions {
+      { &LimiterSettings::thresholdDb, "thresholdDb" },
+      { &LimiterSettings::makeupTargetDb, "makeupTargetDb" },
+      { &LimiterSettings::kneeWidthDb, "kneeWidthDb" },
+      { &LimiterSettings::lookaheadMs, "lookaheadMs" },
+      { &LimiterSettings::releaseMs, "releaseMs" },
+      { &LimiterSettings::showInput, "showInput" },
+      { &LimiterSettings::showOutput, "showOutput" },
+      { &LimiterSettings::showActual, "showActual" },
+      { &LimiterSettings::showTarget, "showTarget" },
+   };
+
+template <typename Struct>
+std::optional<Struct> Deserialize(
+   const std::string& str,
+   const std::vector<SettingDescription<Struct>>& settings)
+{
+   Struct settingStruct;
+
+   const std::regex pattern(R"((\w+)=\"(-?\d+\.?\d*)\")");
```

Comment:
This regex won't recognise settings like:
`1.23e-4`
or
`123.` (without digit after the decimal)
is that ok? just curious

Reply:
Right. It's acceptable insofar as we have unit tests that check that all presets are parsed successfully, so if we introduce one with scientific notation, tests will tell us, and it can be introduced then.
Thank you, though, and thanks unit tests.